packages/babel-plugin-proposal-object-rest-spread/src/index.js

```diff
@@ -180,6 +180,19 @@ export default declare((api, opts) => {
       keyExpression = t.arrayExpression(keys);
     }
 
+    // Hoist definition of excluded keys, so that it's not created each time.
+    if (!t.isProgram(path.scope.block)) {
+      const id = path.scope.generateUidIdentifier();
+
+      file.path.scope.push({
+        id,
+        init: keyExpression,
+        kind: "const",
+      });
+
+      keyExpression = t.cloneNode(id);
```

Comment:
Unfortunately, this won't work for cases with computed keys:

```js
// input
const {
  foo,
  [bar]: baz,
  ...qux
} = obj;

// output
"use strict";

function _objectWithoutPropertiesLoose(source, excluded) {
  if (source == null) return {};
  var target = {};
  var sourceKeys = Object.keys(source);
  var key, i;
  for (i = 0; i < sourceKeys.length; i++) {
    key = sourceKeys[i];
    if (excluded.indexOf(key) >= 0) continue;
    target[key] = source[key];
  }
  return target;
}

function _toPropertyKey(arg) {
  var key = _toPrimitive(arg, "string");
  return typeof key === "symbol" ? key : String(key);
}

function _toPrimitive(input, hint) {
  if (typeof input !== "object" || input === null) return input;
  var prim = input[Symbol.toPrimitive];
  if (prim !== undefined) {
    var res = prim.call(input, hint || "default");
    if (typeof res !== "object") return res;
    throw new TypeError("@@toPrimitive must return a primitive value.");
  }
  return (hint === "string" ? String : Number)(input);
}

var _bar = bar,
  qux = _objectWithoutPropertiesLoose(obj, ["foo", _bar].map(_toPropertyKey));
```

We still have to evaluate that `bar` inline.

Instead, we should be able to call `path.hoist()`, but we'll need to do that after we replace the nodes in the callers of `createObjectRest`.

```js
// line 387
const [inserted] = insertionPath.insertAfter(
  t.variableDeclarator(argument, callExpression),
);
// `arguments.1` is the array expression
inserted.get('arguments.1').hoist()
```

Reply:
For some reason, none of our test seem to hit this path…

Reply:
There's coverage for dynamic props in `transform-destructuring`: `packages/babel-plugin-transform-destructuring/test/fixtures/destructuring/es7-object-rest`.

It's more difficult to do `path.hoist()` on `transform-destructuring`, so I'm keeping the original implementation. Additionally, we'll apply only when `allLiteral` in order to exclude the case you mentioned.